src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -199,6 +200,43 @@ protected Solution SetCurrentSolution(Solution solution)
             }
         }
 
+        /// <summary>
+        /// Applies specified transformation to <see cref="CurrentSolution"/> and updates <see cref="CurrentSolution"/> to the new value.
+        /// </summary>
+        /// <param name="transformation">Solution transformation.</param>
+        /// <param name="oldSolution">Previous instance of the solution.</param>
+        /// <param name="newSolution">Transformed instance of the solution with updated <see cref="Solution.WorkspaceVersion"/>.</param>
+        /// <returns>True if <see cref="CurrentSolution"/> was set to the transformed solution, false if the transformation did not change the solution.</returns>
+        protected bool SetCurrentSolution(Func<Solution, Solution> transformation, out Solution oldSolution, out Solution newSolution)
```

Comment:
It might be useful to add `TArg arg` that's passed to the transformation to avoid closure allocations.